### ISSUE TITLE
AWS EC2 instances left unnamed on failure

### ISSUE THREAD

**Issue by 0x2b3bfa0** [iterative/terraform-provider-iterative#195]
> _Found an interesting riddle to solve. I have an ec2 instance on our account that was running for a few days without activity. It has no name, unlike other cml self-hosted runners, but it was using a keypair with a name that was prefixed with "Iterative-", which is pretty telling that CML did deploy this instance._ [...] _What happened here? Did CML get interrupted in such a way that the instance was started but not named? Are there other ways for CML to skip naming an instance?_ [**(From Discord)**](https://discord.com/channels/485586884165107732/728693131557732403/882662315256864788)

**Comment by 0x2b3bfa0** [comment]
Currently, instances are being created first and then tagged with a [`Name` tag](https://docs.aws.amazon.com/whitepapers/latest/tagging-best-practices/standardize-names-for-aws-resources.html) that gets displayed on the web console as the instance name. Any failure between the creatin and tagging process can leave unnamed instances as described.

https://github.com/iterative/terraform-provider-iterative/blob/4044aee23ea8209ba5244faeaafa8fdbfe92a26e/iterative/aws/provider.go#L270-L286

**Comment by DavidGOrtega** [comment]
Indeed,  as seen here https://github.com/iterative/terraform-provider-iterative/pull/187#issuecomment-916090240 the name is not empty but dashed and the guilty is not even that bit the Id. The main reason it can not be destroyed is because of the Id.

I can see the code here:

```go
func SetId(d *schema.ResourceData) {
	if len(d.Id()) == 0 {
		d.SetId("iterative-" + uid.NewProvider36Size(8).MustGenerate().String())

		if len(d.Get("name").(string)) == 0 {
			d.Set("name", d.Id())
		}
	}
}**
```

For some reason its coming with ```-``` instead of empty

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, https://github.com/iterative/terraform-provider-iterative/pull/187 solves this: now, normal instances are being tagged on creation (i.e. if it exists, it has tags) and spot instance requests are being cancelled after the 10 minute timeout.

I suspect that the issue was that some spot instance requests take more than [10 minutes](https://github.com/aws/aws-sdk-go/blob/23db5a40f4d1f4ede9752ac34c7ed80541a402f3/service/ec2/waiters.go#L1060-L1126) to fulfill, and we were erroring and failing wihtout cancelling the request, so the spot instance would get allocated anytime in the future and out provider wouldn't be running to tag it. 

https://github.com/iterative/terraform-provider-iterative/blob/f7755f7b656c92503beb515adf7ea8fc4dd24921/iterative/aws/provider.go#L231-L242

https://github.com/iterative/terraform-provider-iterative/blob/f7755f7b656c92503beb515adf7ea8fc4dd24921/iterative/aws/provider.go#L271-L286

Since #187, we wait those 10 minutes and then cancel the request before exiting. Not a perfect solution, but good enough for winning the race lottery. 🙃

**Comment by DavidGOrtega** [comment]
Yeah it could be, however... why I was able to reproduce it with the code of #187 applied? @0x2b3bfa0

**Comment by DavidGOrtega** [comment]
Anyway lets close this once #187 is published!

**Comment by DavidGOrtega** [comment]
My other question would be.... why a dash? Its an AWS thingy?

**Comment by 0x2b3bfa0** [comment]
> _My other question would be.... why a dash? Its an AWS thingy?_

Yes, that's what AWS displays when there is no `Name` tag for the resource. 🙃

**Comment by 0x2b3bfa0** [comment]
Closed with #187
